python/pyarrow/tests/test_table.py

```diff
@@ -2297,6 +2297,49 @@ def test_table_join_many_columns():
     })
 
 
+@pytest.mark.dataset
+def test_table_join_asof():
+    t1 = pa.Table.from_pydict({
+        "colA": [1, 1, 5, 6, 7],
+        "col2": ["a", "b", "a", "b", "f"]
+    })
+
+    t2 = pa.Table.from_pydict({
+        "colB": [2, 9, 15],
+        "col3": ["a", "b", "g"],
+        "colC": [1., 3., 5.]
+    })
+
+    r = t1.join_asof(t2, "colA", "col2", 1, "colB", "col3")
```

Comment:
Some additional test case ideas to ensure good coverage:

- A test where the left/right column names are the same, so you can rely on not having to specify right_on/by
- A test where the `by` keys is a list of columns instead of a single one

Reply:
> A test where the left/right column names are the same, so you can rely on not having to specify right_on/by

This is now tested.

> A test where the by keys is a list of columns instead of a single one

This is now tested.

> and what happens if passing an empty list?

It seems like it just doesn't perform the join over any partitions - this is also now tested.